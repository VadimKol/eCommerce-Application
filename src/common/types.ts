import type { BaseAddress, CustomFieldsDraft, StoreResourceIdentifier } from '@commercetools/platform-sdk';

export interface AuthContextInterface {
  isAuthenticated: boolean;
  handleLogin: () => void;
  handleLogout: () => void;
}

export type CategoriesContextType = {
  categories: CategoriesData | null;
  loading: boolean;
  error: Error | null;
};

export interface Credentials {
  clientId: string;
  clientSecret: string;
}

export interface User {
  username: string;
  password: string;
}

export interface GeekShopCustomerDraft {
  /**
   *	Email address of the Customer that is [unique](/../api/customers-overview#customer-uniqueness) for an entire Project or Store the Customer is assigned to.
   *	It is the mandatory unique identifier of a Customer.
   *
   *
   */
  readonly email: string;
  /**
   *	Password of the Customer.
   *
   *
   */
  readonly password: string;
  /**
   *	Given name (first name) of the Customer.
   *
   *
   */
  readonly firstName?: string;
  /**
   *	Family name (last name) of the Customer.
   *
   *
   */
  readonly lastName?: string;
  /**
   *	Middle name of the Customer.
   *
   *
   */
  readonly middleName?: string;
  /**
   *	Title of the Customer, for example, 'Dr.'.
   *
   *
   */
  readonly title?: string;
  /**
   *	Salutation of the Customer, for example, 'Mr.' or 'Mrs.'.
   *
   *
   */
  readonly salutation?: string;
  /**
   *	Date of birth of the Customer.
   *
   *
   */
  readonly dateOfBirth?: string;
  /**
   *	Company name of the Customer.
   *
   *
   */
  readonly companyName?: string;
  /**
   *	Individual VAT ID of the Customer.
   *
   *
   */
  readonly vatId?: string;
  /**
   *	Addresses of the Customer.
   *
   *
   */
  readonly addresses?: BaseAddress[];
  /**
   *	Indexes of entries in `addresses` to set as shipping addresses.
   *	The `shippingAddressIds` of the [Customer](ctp:api:type:Customer) will be replaced by these addresses.
   *
   *
   */
  readonly shippingAddresses?: number[];
  /**
   *	Index of the address in the `addresses` array to use as the default shipping address.
   *	The `defaultShippingAddressId` of the Customer will be set to the `id` of that address.
   *
   *
   */
  readonly defaultShippingAddress?: number;
  /**
   *	Indexes of entries in `addresses` to set as billing addresses.
   *	The `billingAddressIds` of the [Customer](ctp:api:type:Customer) will be replaced by these addresses.
   *
   *
   */
  readonly billingAddresses?: number[];
  /**
   *	Index of the address in the `addresses` array to use as the default billing address.
   *	The `defaultBillingAddressId` of the Customer will be set to the `id` of that address.
   *
   *
   */
  readonly defaultBillingAddress?: number;
  /**
   *	Custom Fields for the Customer.
   *
   *
   */
  readonly custom?: CustomFieldsDraft;
  /**
   *	Preferred language of the Customer. Must be one of the languages supported by the [Project](ctp:api:type:Project).
   *
   *
   */
  readonly locale?: string;
  /**
   *	Sets the [Stores](ctp:api:type:Store) for the Customer.
   *
   *
   */
  readonly stores?: StoreResourceIdentifier[];
}

export type ProductDetails = {
  name?: string;
  description?: string;
  price?: string | null;
  discountedPrice?: string | null;
  currency?: string | null;
  images?: string[];
  attributes?: { name: string; value: string | number | string[] }[];
  availability?: {
    isOnStock?: boolean | null;
    availableQuantity?: number | null;
  };
};

export type Product = {
  id: string;
  name: string;
  description: string;
  slug: string;
  key: string;
  sku: string;
  quantity: number;
  price: string;
  images: string[];
  categoryId: string;
  slugCategory: string;
  keyCategory: string;
  subcategoryId: string;
  slugSubCategory: string;
  keySubCategory: string;
};

export type Subcategory = {
  id: string;
  name: string;
  slug: string;
  key: string;
  products: Product[];
};

export type Category = {
  id: string;
  name: string;
  slug: string;
  key: string;
  subcategories: Subcategory[];
};

export type CategoriesData = Category[];

export type UrlParams = {
  categoryName?: string;
  subcategoryName?: string;
  productName?: string;
};

export type LoaderData = {
  category?: Category;
  subcategory?: Subcategory;
  product?: Product;
};
