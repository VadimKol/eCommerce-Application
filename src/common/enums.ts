export enum NavigationPaths {
  HOME = '/',
  CATALOG = '/catalog',
  ABOUT = '/about',
}

export enum ActionPaths {
  LOGIN = '/login',
  REGISTER = '/register',
  PROFILE = '/profile',
}
