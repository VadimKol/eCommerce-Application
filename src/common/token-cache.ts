import type { TokenCache, TokenStore } from '@commercetools/sdk-client-v2';

export class CustomTokenCache implements TokenCache {
  private tokenStore: TokenStore = {
    token: localStorage.getItem('geek-shop-token') ? (localStorage.getItem('geek-shop-token') as string) : '',
    expirationTime: 0,
    refreshToken: '',
  };

  public set(tokenStore: TokenStore): void {
    this.tokenStore = tokenStore;
  }

  public get(): TokenStore {
    return this.tokenStore;
  }
}
