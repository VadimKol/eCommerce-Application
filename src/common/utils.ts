import type { CatalogAction, CatalogState } from './types';

export const calculateAge = (birthDate: string): boolean => {
  const today = new Date();
  const birth = new Date(birthDate);
  let age = today.getFullYear() - birth.getFullYear();
  const monthDifference = today.getMonth() - birth.getMonth();
  if (monthDifference < 0 || (monthDifference === 0 && today.getDate() < birth.getDate())) {
    age -= 1;
  }
  return age >= 13;
};

export const sortingTypes = ['Alphabetically, A-Z', 'Alphabetically, Z-A', 'Price, low to high', 'Price, high to low'];

export const fandoms = [
  'Genshin',
  'Disney',
  'Cyberpunk',
  'Batman',
  'Doctor Who',
  'Star Wars',
  'Beetlejuice',
  'Marvel',
  'One Piece',
  'Harry Potter',
  'Stranger Things',
  'Rick and Morty',
  'Supernatural',
];

export const countriesFilter = ['China', 'France', 'Germany', 'India', 'Italy', 'Korea', 'US'];
export const materials = ['Aluminum', 'Cardboard', 'Ceramic', 'Coir', 'Paper', 'Plastic', 'Polyester'];
export const countries = [
  {
    title: 'United States',
    code: 'US',
    errorMsg: 'It must be 5 digits or 5+4 digits',
    postcode: /^\d{5}(-\d{4})?$/,
  },
  {
    title: 'Belarus',
    code: 'BY',
    errorMsg: 'It must be exactly 6 digits.',
    postcode: /^\d{6}$/,
  },
  {
    title: 'Russia',
    code: 'RU',
    errorMsg: 'It must be exactly 6 digits.',
    postcode: /^\d{6}$/,
  },
];

export const DEFAULT_LOCALE = 'en-US';
export const QUERY_LIMIT = 12;
export const PRICE_FILTER_MIN = 1;
export const PRICE_FILTER_MAX = 100;
export const REFRESH_TOKEN_EXPIRATION_DAYS = 150;
export const Discounts = [
  {
    name: 'GEEK-SHOP',
    desc: '-10% cart',
  },
  {
    name: 'COSPLAY',
    desc: '-10% category',
  },
];

export function getCheckboxFilter(checkboxes: boolean[], attribute: string): string {
  const filter: string[] = [];
  let constantsFilter: string[] = [];

  switch (attribute) {
    case 'Fandom': {
      constantsFilter = fandoms;
      break;
    }
    case 'Country': {
      constantsFilter = countriesFilter;
      break;
    }
    case 'Material': {
      constantsFilter = materials;
      break;
    }
    default:
      break;
  }

  checkboxes.forEach((checkbox, index) => {
    constantsFilter.forEach((element, i) => {
      if (index === i && checkbox) {
        filter.push(`"${element}"`);
      }
    });
  });

  return filter.length > 0 ? `variants.attributes.${attribute}:${filter.join(',')}` : '';
}

export function getSort(sortType: string): string | undefined {
  switch (sortType) {
    case 'Alphabetically, A-Z':
      return `name.${DEFAULT_LOCALE} asc`;
    case 'Alphabetically, Z-A':
      return `name.${DEFAULT_LOCALE} desc`;
    case 'Price, low to high':
      return 'price asc';
    case 'Price, high to low':
      return 'price desc';
    default:
      return undefined;
  }
}

export class StatusError extends Error {
  public statusCode: number;

  constructor(message: string, status: number) {
    super(message);
    this.statusCode = status;
  }
}

export const initialState: CatalogState = {
  products: [],
  page: 0,
  sortType: 'Sort',
  search: '',
  priceFilter: [PRICE_FILTER_MIN, PRICE_FILTER_MAX],
  franchises: Array(fandoms.length).fill(false),
  countriesF: Array(countriesFilter.length).fill(false),
  materialsF: Array(materials.length).fill(false),
  loadingProducts: true,
  categories: {},
};

export function reducerCatalog(
  state: CatalogState,
  {
    type,
    products = [],
    page = 0,
    sortType = 'Sort',
    search = '',
    priceFilter = [PRICE_FILTER_MIN, PRICE_FILTER_MAX],
    franchises = Array(fandoms.length).fill(false),
    countriesF = Array(countriesFilter.length).fill(false),
    materialsF = Array(materials.length).fill(false),
    loadingProducts = true,
    categories = {},
  }: CatalogAction,
): CatalogState {
  switch (type) {
    case 'SET_CATEGORIES':
      return {
        ...state,
        categories,
        page,
        sortType,
        priceFilter,
        franchises,
        countriesF,
        materialsF,
        search,
        loadingProducts,
      };
    case 'SET_PRODUCTS':
      return { ...state, products };
    case 'SET_LOADING_PRODUCTS':
      return { ...state, loadingProducts };
    case 'SET_SEARCH':
      return { ...state, search, page, loadingProducts };
    case 'SET_SORT_TYPE':
      return { ...state, sortType, loadingProducts };
    case 'SET_PRICE':
      return { ...state, priceFilter, page, loadingProducts };
    case 'SET_FRANCHISES':
      return { ...state, franchises, page, loadingProducts };
    case 'SET_COUNTRIES':
      return { ...state, countriesF, page, loadingProducts };
    case 'SET_MATERIALS':
      return { ...state, materialsF, page, loadingProducts };
    case 'RESET_FILTERS':
      return { ...state, priceFilter, franchises, countriesF, materialsF, page, loadingProducts };
    case 'SET_PAGE':
      return { ...state, page, loadingProducts };
    default:
      return state;
  }
}

export const capitalizeFirstLetter = (string: string): string => string.charAt(0).toUpperCase() + string.slice(1);
