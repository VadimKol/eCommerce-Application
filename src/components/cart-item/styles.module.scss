.remove {
  flex-shrink: 0;
  align-self: center;
  width: 25px;
  height: 25px;
  padding: 0;
  background-color: transparent;
  background-image: url("/src/assets/images/delete.png");
  background-size: cover;
  border: none;
  transition: transform 0.5s;

  @media (hover: hover) {
    &:hover:not(:disabled) {
      cursor: pointer;
      transform: scale(1.1);
    }
  }

  &:disabled {
    cursor: default;
  }

  @media (width <= 570px) {
    width: 35px;
    height: 35px;
  }
}

.image_box {
  display: flex;
  flex-shrink: 0;
  justify-content: center;
  align-items: center;
  align-self: center;
  overflow: hidden;
  width: 200px;
  height: 200px;
  border-radius: 6px;

  @media (width <= 1300px) {
    width: 140px;
    height: 140px;
  }

  @media (width <= 570px) {
    order: 1;
    width: 200px;
    height: 200px;
  }
}

.product_img {
  max-width: 100%;
  max-height: 100%;
}

.product_desc {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  gap: 10px;
}

.product_name {
  flex-grow: 1;
  margin: 0;
  font-size: 1.8rem;
  line-height: 1.2;

  @media (width <= 1300px) {
    font-size: 1.2rem;
  }

  @media (width <= 570px) {
    margin: 0 0 1.2rem;
    text-align: center;
  }
}

.product_details {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  gap: 5px;

  @media (width <= 570px) {
    align-self: center;
  }
}

.quantity_box {
  font-size: 1.3rem;

  @media (width <= 1300px) {
    font-size: 0.91rem;
  }
}

.quantity_btn,
.quantity_btn_loading {
  width: 1.25em;
  padding: 0;
  font-family: var(--primary-font);
  font-size: 1em;
  color: var(--primary-color);
  background-color: var(--background-color);
  border: 1px solid var(--primary-color);
  border-radius: 3px;
  transition: all 0.5s;


  @media (hover: hover) {
    &:hover {
      color: var(--background-color);
      background-color: var(--primary-color);
      cursor: pointer;
    }
  }

  @media (width <= 1300px) {
    width: 20px;
  }
}

.quantity_btn {
  &:disabled {
    color: var(--background-color);
    background-color: var(--disabled-color);
    border: 1px solid var(--disabled-color);
    cursor: default;
  }
}

.quantity {
  display: inline-block;
  margin: 0 0.48em;
}

.available_quantity,
.quantity {
  font-weight: 600;
}

.max_quantity_desc {
  margin-left: 0.48em;
}

.total_block {
  display: flex;
  flex-grow: 1;
  align-items: flex-end;
  font-size: 1.8rem;

  @media (width <= 1300px) {
    font-size: 1.2rem;
  }
}

.price_block {
  display: flex;
  margin: 0;
  font-size: 1.3rem;

  @media (width <= 1300px) {
    font-size: 0.91rem;
  }
}

.price {
  display: flex;
  align-items: center;
  font-weight: 600;
}

.product_discount {
  margin-left: 0.65em;
  font-size: 0.8em;
  line-height: 1;
  text-decoration: line-through;
  color: var(--error-color);
}

.text {
  display: inline-block;
  width: 9.5rem;

  @media (width <= 1300px) {
    width: 6.5rem;
  }
}
