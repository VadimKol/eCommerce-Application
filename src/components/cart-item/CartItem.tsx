import { type LineItem } from '@commercetools/platform-sdk';
import { useState } from 'react';
import { toast } from 'react-toastify';

import { DEFAULT_LOCALE } from '@/common/utils';
import { useCart } from '@/hooks/useCart';

import styles from './styles.module.scss';

export function CartItem({ product }: { product: LineItem }): JSX.Element {
  const { removeItemFromCart, changeItemQuantityFromCart } = useCart();
  const [isRemoving, setIsRemoving] = useState(false);
  const [isChangingQuantity, setIsChangingQuantity] = useState(false);

  const price = (product.price.value.centAmount / 100).toFixed(2);
  let discount = product.price.discounted ? (product.price.discounted.value.centAmount / 100).toFixed(2) : '';
  if (product.discountedPricePerQuantity.length > 0) {
    discount = product.discountedPricePerQuantity[0]
      ? (product.discountedPricePerQuantity[0].discountedPrice.value.centAmount / 100).toFixed(2)
      : discount;
  }
  const totalPrice = (Number(price) * product.quantity).toFixed(2);
  const discountOnTotalPrice = (Number(discount) * product.quantity).toFixed(2);

  return (
    <>
      <button
        type="button"
        aria-label="remove"
        className={styles.remove}
        onClick={() => {
          setIsRemoving(true);
          removeItemFromCart(product.productId)
            .then(() => toast('Successfully removed', { type: 'success' }))
            .catch(() => toast('Failed to remove', { type: 'error' }))
            .finally(() => setIsRemoving(false));
        }}
        disabled={isRemoving}
      />
      <div className={styles.image_box}>
        <img
          className={styles.product_img}
          src={product.variant.images ? product.variant.images[0]?.url : ''}
          alt="Product"
        />
      </div>
      <div className={styles.product_desc}>
        <p className={styles.product_name}>{product.name[DEFAULT_LOCALE]}</p>
        <div className={styles.product_details}>
          <div className={styles.quantity_box}>
            <span className={styles.text}>Quantity:</span>
            <button
              type="button"
              aria-label="minus"
              className={isChangingQuantity ? styles.quantity_btn_loading : styles.quantity_btn}
              onClick={() => {
                setIsChangingQuantity(true);
                changeItemQuantityFromCart(product.productId, product.quantity - 1)
                  .catch(() => toast('Failed to change quantity', { type: 'error' }))
                  .finally(() => setIsChangingQuantity(false));
              }}
              disabled={isChangingQuantity}
            >
              -
            </button>
            <span className={styles.quantity}>{product.quantity}</span>
            <button
              type="button"
              aria-label="plus"
              className={isChangingQuantity ? styles.quantity_btn_loading : styles.quantity_btn}
              onClick={() => {
                setIsChangingQuantity(true);
                changeItemQuantityFromCart(product.productId, product.quantity + 1)
                  .catch(() => toast('Failed to change quantity', { type: 'error' }))
                  .finally(() => setIsChangingQuantity(false));
              }}
              disabled={product.quantity === product.variant.availability?.availableQuantity || isChangingQuantity}
            >
              +
            </button>
            <span className={styles.max_quantity_desc}>
              out of{' '}
              <span className={styles.available_quantity}>{product.variant.availability?.availableQuantity}</span>
            </span>
          </div>
          <p className={styles.price_block}>
            <span className={styles.text}>Price per item:</span>
            <div className={styles.price}>
              {discount && <span>${discount}</span>}
              <span className={discount && styles.product_discount}>${price}</span>
            </div>
          </p>
          <p className={styles.total_block}>
            <span className={styles.text}>Items cost:</span>
            <div className={styles.price}>
              {discount && <span>${discountOnTotalPrice}</span>}
              <span className={discount && styles.product_discount}>${totalPrice}</span>
            </div>
          </p>
        </div>
      </div>
    </>
  );
}
