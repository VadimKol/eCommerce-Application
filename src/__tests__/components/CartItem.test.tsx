import { type LineItem } from '@commercetools/platform-sdk';
import { render } from '@testing-library/react';

import { DEFAULT_LOCALE } from '@/common/utils';
import { CartItem } from '@/components/cart-item/CartItem';

it('renders correctly', () => {
  const product: LineItem = {
    id: '',
    productId: '',
    name: {
      [DEFAULT_LOCALE]: '',
    },
    productType: {
      typeId: 'product-type',
      id: '',
    },
    variant: {
      id: 0,
    },
    price: {
      id: '',
      value: {
        type: 'centPrecision',
        centAmount: 0,
        currencyCode: 'USD',
        fractionDigits: 2,
      },
    },
    quantity: 0,
    totalPrice: {
      type: 'centPrecision',
      centAmount: 0,
      currencyCode: 'USD',
      fractionDigits: 2,
    },
    discountedPricePerQuantity: [],
    taxedPricePortions: [],
    state: [],
    perMethodTaxRate: [],
    priceMode: 'Platform',
    lineItemMode: 'Standard',
  };
  const { container } = render(<CartItem product={product} />);
  expect(container).toMatchSnapshot();
});
