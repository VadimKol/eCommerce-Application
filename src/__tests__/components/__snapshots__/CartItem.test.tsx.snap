// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <button
    aria-label="remove"
    class="remove"
    type="button"
  />
  <div
    class="image_box"
  >
    <img
      alt="Product"
      class="product_img"
      src=""
    />
  </div>
  <div
    class="product_desc"
  >
    <p
      class="product_name"
    />
    <div
      class="product_details"
    >
      <div
        class="quantity_box"
      >
        <span
          class="text"
        >
          Quantity:
        </span>
        <button
          aria-label="minus"
          class="quantity_btn"
          type="button"
        >
          -
        </button>
        <span
          class="quantity"
        >
          0
        </span>
        <button
          aria-label="plus"
          class="quantity_btn"
          type="button"
        >
          +
        </button>
        <span
          class="max_quantity_desc"
        >
          out of
           
          <span
            class="available_quantity"
          />
        </span>
      </div>
      <div
        class="price_block"
      >
        <span
          class="text"
        >
          Price per item:
        </span>
        <div
          class="price"
        >
          <span
            class=""
          >
            $
            0.00
          </span>
        </div>
      </div>
      <div
        class="total_block"
      >
        <span
          class="text"
        >
          Items cost:
        </span>
        <div
          class="price"
        >
          <span
            class=""
          >
            $
            0.00
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
